### ISSUE TITLE
Enable 'Show source' for events

### ISSUE THREAD

**Issue by jplatte** [matrix-org/matrix-rust-sdk#187]
Continuing the discussion from https://github.com/ruma/ruma/issues/442#issuecomment-805738519:

> I think there are three distinct problems here:
> 
> 1. [...]
> 2. [...]
> 3. SDK users get deserialized events and thus they might not be able to show the correct source of an event.
> 
> [...]
> 
> Now for 3. to be able to show the source of the event using `Raw<_>` we would have to give users each event in `Raw<_>` form, this is problematic since we already need to deserialize the events in the SDK to update and change our state, and giving out `Raw<_>` turned out to be somewhat of a footgun. [...]

I get that you have to deserialize anyways. But deserializing doesn't consume the `Raw<_>` value. I just took a look at how `EventHandler` is used and saw that many / all methods have a `room` parameter. From that, I have a suggestion: Change

```rust
async fn on_room_message(&self, room: Room, event: &SyncMessageEvent<MessageEventContent>)
```

to

```rust
async fn on_room_message(&self, event: &SyncMessageEvent<MessageEventContent>, ctx: SyncMessageCtx)
```

where `SyncMessageCtx` is a struct containing the `room`, `encryption_info` and `raw: Box<RawJsonValue>`.

This means you can give users access to encryption info and raw JSON easily, without a huge number of new types like #136 does and without making the trait method signatures more complex.

> > And then you still wouldn't have the original representation of the event for things like Show source since we intentionally throw away information, the classical example being None representing either an absent field or a null (but there's also more pronounced things like default power levels not being re-serialized).
> 
> Is this a big problem? Do we have an estimate how many events do this? I would rather sacrifice the complete correctness here to get an easier to use API.

To me, it's a big problem. Correctness is one of the core goals of Ruma. And in particular for a feature like 'Show source' whose main use case is debugging, being precise seems very important to me.

**Comment by poljar** [comment]
> where `SyncMessageCtx` is a struct containing the `room`, `encryption_info` and `raw: Box<RawJsonValue>`.
> 
> This means you can give users access to encryption info and raw JSON easily, without a huge number of new types like #136 does and without making the trait method signatures more complex.

On the surface this seems to be a great suggestions, sadly the event handler isn't the only way to consume events, anyone that want's to consume the events in bulk will want to consume whole responses, e.g. you don't want to push something like [matrix-recorder](https://gitlab.com/argit/matrix-recorder) into using a transaction per event.

Pantalaimon and anything that will use the base crate over FFI will want to do the same. The event handler is mostly useful for clients and bots.

Since we have already a custom `SyncResponse` type we could add a similar context that will hold the Raw version of the event in a separate field of the response as well, but I think it'll be problematic to explain to people why they can't just take the event, serialize it and store it in their SQLite database.

Why do they need to use one type to analyze the event and another one to serialize it. It might make more sense to just store the `Raw` event inside the deserialized one and use the `Raw` one for serialization. This of course means that we're doubling our memory footprint, though the same is true if we go the context way.

**Comment by jplatte** [comment]
> Since we have already a custom `SyncResponse` type we could add a similar context that will hold the Raw version of the event in a separate field of the response as well, but I think it'll be problematic to explain to people why they can't just take the event, serialize it and store it in their SQLite database.

I'm not sure it would be so hard to explain.

> It might make more sense to just store the `Raw` event inside the deserialized one and use the `Raw` one for serialization.

This would mean having separate event types for everything in matrix-sdk. I think matrix-sdk is more comprehensible when it doesn't duplicate half of Ruma.

> This of course means that we're doubling our memory footprint, though the same is true if we go the context way.

I was thinking you construct the context before passing it to the event handler, rather than constructing the context for every event in a sync response at once and keeping those all in memory at the same time.

> On the surface this seems to be a great suggestions, sadly the event handler isn't the only way to consume events, anyone that want's to consume the events in bulk will want to consume whole responses

Aren't there ways in which the event handler could be made to better suit those cases? It seems like you currently have two entirely separate interfaces in one library, and that leads to problems like this.

**Comment by jplatte** [comment]
Another discussion happened about this and related concerns on Matrix, starting at https://matrix.to/#/!iYnZafYUoXkeVPOSQh:matrix.org/$KMnnkqTnwIBspGjENeVLj91lBKmE8hclGCQjnO182_c?via=matrix.org&via=gnome.org&via=tedomum.net, resulting in PR #209. There's still some work to do after that, but I think there is now a clear(er?) path towards resolving this.

**Comment by poljar** [comment]
This has been solved by #216.
